src/components/orator/js/Orator.js

```diff
@@ -0,0 +1,571 @@
+     */
+    fluid.orator.domReader.readFromDOM = function (that, elm) {
+        elm = $(elm);
+
+        // only execute if there are nodes to read from
+        if (elm.length) {
+            var parsedFromElm = fluid.orator.domReader.parse(elm[0]);
+            that.parseQueue = parsedFromElm;
+            that.events.onReadFromDOM.fire(parsedFromElm);
+            that.tts.queueSpeech(fluid.orator.domReader.parsedToString(parsedFromElm));
+        }
+    };
+
+    /**
+     * Returns the index of the closest data point from the parseQueue based on the boundary provided.
+     *
+     * @param {Array} parseQueue - An array data points generated from parsing a DOM structure
+     * @param {Integer} currentIndex - The index into the paraseQueue to start searching from. The currentIndex will be
+     *                                constrained to the bounds of the parseQueue.
+     * @param {Integer} boundary - The boundary value used to compare against the blockIndex of the parsed data points.
+     *
+     * @return {Integer|undefined} - Will return the index of the closest data point in the parseQueue. If the boundary
+     *                               cannot be located within the parseQueue, `undefined` is returned.
+     */
+    fluid.orator.domReader.getClosestIndex = function (parseQueue, currentIndex, boundary) {
+        var maxIdx  = Math.max(parseQueue.length - 1, 0);
+        currentIndex = Math.max(Math.min(currentIndex, maxIdx), 0);
+
+        var nextIdx = currentIndex + 1;
+        var prevIdx = currentIndex - 1;
+
+        var currentBlockIndex = parseQueue[currentIndex].blockIndex;
+        var maxBoundary = parseQueue[maxIdx].blockIndex + parseQueue[maxIdx].word.length;
+
+
+        if (!fluid.isValue(boundary) || boundary < 0 || boundary > maxBoundary ) {
+            return undefined;
+        }
+
+        if (currentBlockIndex > boundary) {
+            return fluid.orator.domReader.getClosestIndex(parseQueue, prevIdx, boundary);
+        }
+
+        var isInNextBound = parseQueue[nextIdx] ? boundary < parseQueue[nextIdx].blockIndex : boundary <= maxBoundary;
+
+        if (currentBlockIndex === boundary || (currentIndex <= maxIdx && isInNextBound)) {
+            return currentIndex;
+        }
+
+        return fluid.orator.domReader.getClosestIndex(parseQueue, nextIdx, boundary);
+    };
+
+    /**
+     * Highlights text from the parseQueue according to the specified boundary. Highlights are performed by wrapping
+     * the appropriate text in the markup specified by `that.options.markup.highlight`.
+     *
+     * @param {Component} that - the component
+     * @param {Integer} boundary - the boundary point used to find the text to highlight. Typically this is the
+     *                             utterance boundary returned from the utteranceOnBoundary event.
+     */
+    fluid.orator.domReader.highlight = function (that, boundary) {
+        that.removeHighlight();
+
+        if (that.parseQueue.length) {
```

Comment:
This should be a modelListener to something rather than an arbitrarily scheduled invoker

Reply:
should probably modelise the boundary index as well.

Reply:
May be able to setup a modelRelay to a closestIndex model value which has a relay rule that contains the functionality currently in getClosestIndex. Then an updateHighlight method could be used to highlight the appropriate portion using this index.